### PR TITLE
[Testnet] NFT Contract -  Certik Reviews

### PR DESCRIPTION
**Pull request by YakupIpek**
# Non-Fungible Token Contract

**Compiler Version**
```
```
**Contract Hash**
```
1c4dafabf65c2e28e5ed451d69947ef4e5a8c3ad5d113f24537a5cf561db4693
```

**Contract Byte Code**
```
4D5A90000300000004000000FFFF0000B800000000000000400000000000000000000000000000000000000000000000000000000000000000000000800000000E1FBA0E00B409CD21B8014CCD21546869732070726F6772616D2063616E6E6F742062652072756E20696E20444F53206D6F64652E0D0D0A2400000000000000504500004C010200DFEAAD8A0000000000000000E00022200B013000002000000002000000000000A23F0000002000000040000000000010002000000002000004000000000000000400000000000000006000000002000000000000030040850000100000100000000010000010000000000000100000000000000000000000503F00004F000000000000000000000000000000000000000000000000000000004000000C000000343F00001C0000000000000000000000000000000000000000000000000000000000000000000000000000000000000000200000080000000000000000000000082000004800000000000000000000002E74657874000000A81F0000002000000020000000020000000000000000000000000000200000602E72656C6F6300000C000000004000000002000000220000000000000000000000000000400000420000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000843F00000000000048000000020005006C280000C816000001000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000007202280500000A7201000070038C0E000001280600000A6F0700000A2A7A02280500000A720100007003B88C0E000001280600000A046F0800000A2A46722F000070038C06000001280600000A2A4E02280500000A020328030000066F0900000A2A5202280500000A02032803000006046F0A00000A2A46724B000070038C06000001280600000A2A4E02280500000A020328060000066F0900000A2A5202280500000A02032806000006046F0A00000A2A7202280500000A726D000070038C07000001280600000A6F0B00000A2A7602280500000A726D000070038C07000001280600000A046F0C00000A2A8A02280500000A7285000070038C07000001048C07000001280D00000A6F0700000A2A8E02280500000A7285000070038C07000001048C07000001280D00000A056F0800000A2A4602280500000A72B50000706F0900000A2A4A02280500000A72B5000070036F0A00000A2A4602280500000A72C10000706F0900000A2A4A02280500000A72C1000070036F0A00000A2A4602280500000A72DB0000706F0E00000A2A4A02280500000A72DB000070036F0F00000A2A4602280500000A72E50000706F0E00000A2A4A02280500000A72E5000070036F0F00000A2A7202280500000A72F3000070038C06000001280600000A6F0E00000A2A7602280500000A72F3000070038C06000001280600000A046F0F00000A2A4602280500000A72030100706F0700000A2A4A02280500000A7203010070036F0800000A2A4602280500000A72250100706F0B00000A2A4A02280500000A7225010070036F0C00000A2A0000001330030057000000000000000203281000000A021717280200000602181728020000060219162802000006021A172802000006021B16280200000602042812000006020528140000060202281100000A6F1200000A280E000006020E0428180000062A32020304050E0428280000062A4202030405168D1200000128280000062A000000133004003A0000000100001102042839000006020528040000060A02062839000006020605282F000006020603281300000A7243010070281400000A0206040528250000062A0000133004004400000001000011020428040000060A020628390000060206282D000006020306281500000A729101007072E1010070280600000A281400000A020403280800000602060304282B0000062A9E0202281100000A6F1200000A0304280C0000060202281100000A6F1200000A0304282C0000062A3E02032839000006020328090000062A133002001100000001000011020328040000060A02062839000006062A000000133002001700000001000011020328040000060A02062839000006020328070000062A26020304280B0000062A86020528290000060203052826000006020405282700000602030405282A0000062A00133004002800000002000011020328090000060A02030617281600000A281700000A280A00000602047E1800000A28050000062A1330040024000000020000110204032805000006020328090000060A02030617281600000A281900000A280A0000062A5602030405281E000006020304050E0428360000062A82020328070000067E1800000A281500000A2C0C02037E1800000A28080000062A00133003002800000003000011021200FE15040000021200037D070000041200047D080000041200057D0900000406280100002B2A133003002800000004000011021200FE15050000021200037D0A0000041200047D0B0000041200057D0C00000406280200002B2A133003002800000005000011021200FE15060000021200037D0D0000041200047D0E0000041200057D0F00000406280300002B2A4E020203282E00000672F3010070281400000A2AA20302281100000A6F1200000A281300000A2D13020302281100000A6F1200000A280B0000062A172ABE020203282E0000062D190204280700000602281100000A6F1200000A281300000A2B01177237020070281400000A2A00000013300300320000000600001102283200000602032810000006021200FE1508000002120002280D0000067D120000041200037D1300000406280400002B2A0000133003005C0000000700001102280F0000060A020602281100000A6F1200000A281300000A728F020070281400000A02280D0000060B0206280E000006027E1800000A2810000006021202FE15070000021202077D100000041202067D1100000408280500002B2A8A0202281100000A6F1200000A02280D000006281300000A7201030070281400000A2A001330040026000000080000110202281900000617281600000A281900000A250B281A000006070A020306042835000006062A00001330050034000000080000110202281900000617281600000A281900000A250B281A000006070A020306042835000006027E1800000A0306052836000006062ACE0228170000062C060228320000060203283900000602030428270000060204052816000006027E1800000A0304282A0000062A13300800660000000900001102280500000A046F1B00000A2C570204166A72350300701A8D10000001251602281100000A6F1200000A8C07000001A22517038C07000001A22518058C06000001A225190E04A2166A281C00000A0A02066F1D00000AA513000001726B030070281400000A2A0000133004004900000001000011020328040000060A020602281100000A6F1200000A281300000A72BB030070281400000A020328290000060206032826000006020314281600000602067E1800000A03282A0000062A22020328150000062A5E02037E1800000A281500000A7205040070281400000A2A000042534A4201000100000000000C00000076342E302E33303331390000000005006C00000014090000237E000080090000EC06000023537472696E6773000000006C1000004004000023555300AC140000100000002347554944000000BC1400000C02000023426C6F620000000000000002000001571DA201090A000000FA01330016000001000000130000000800000013000000390000005C0000001D000000050000000F0000000900000001000000060000000C000000010000000200000006000000050000000000700301000000000006009E01BF050600CD01BF0506008A0195050F00DF0500000A0045061C060A0016001C060A000C061C060A005F011C060A0066061C060600CC03AF0306004001AF030A00BE011C060A0073011C0606000100AF0306002D02AF0306005306AF030A00F8001C060600EB01AF030600D103AF03000000001E00000000000100010001001000FF03000015000100010003010000C0000000290001003A000A011000C20200002D0007003A000A011000A40200002D000A003A000A011000B00200002D000D003A000A0110006F0200002D0010003A000A011000860200002D0012003A00060653000E015680CF0011015680FE03110156803D05110156805B00110156800D0111010600C203890006003C04890006009800150106000B0589000600AB00890006009800150106000B0589000600860589000600AB0019010600E40489000600FF0489000600F20489000600B20489005020000000008600D0001C0101006D20000000008100A805210102008C200000000081009906280104009E20000000008100CA042E010500B220000000008100D70435010600C720000000008100860628010800D92000000000810008032E010900ED20000000008100180335010A000221000000008100E2003D010C001F21000000008100ED0044010D003D2100000000810057054C010F0060210000000081006A0554011100842100000000860879045D001400962100000000810883045D011400A9210000000086088D045D001500BB210000000081089E045D011500CE21000000008608290163011600E021000000008108320167011600F32100000000860882036301170005220000000081088D03670117001822000000008100330028011800352200000000810043006C0119005322000000008108340273011B006522000000008108490277011B00782200000000810817057C011C008A220000000081082A0581011C00A0220000000086188F0587011D000323000000008600B603900121001023000000008600B6039C0125002423000000008600BA039C0128006C23000000008600060244012B00BC230000000086005903A6012D00E4230000000086000E023D012F00F42300000000860018022E0130001424000000008600A8002E013100372400000000860036034C0132004124000000008100EB029C01340064240000000081001004440137009824000000008100D90344013900C824000000008100D20290013B00DE24000000008100280381013F00002500000000810061049C0140003425000000008100FC029C01430068250000000081004703540146009C250000000081004A015D014900B0250000000081007D05D6004A00D9250000000081006D0444014B000C26000000008600AF045D014D004C26000000008600470406004E00B426000000008100A90606004E00D8260000000086007A06AD014E000C270000000086007606B50150004C270000000081007A06BF0153008027000000008100E20390015600F427000000008600370481015A0049280000000086004A0028015B005228000000008100D3065D015C00000001002700000001008C0000000200000200000100BD0000000100BD0000000100BD0000000200000200000100BD0000000100BD0000000100BD0000000200000200000100140600000100140600000200000200000100110500000200EE0500000100110500000200EE0500000300000200000100000200000100000200000100000200000100000200000100A00000000100A00000000200CE02000001000002000001000002000001008401000002003B01000003009803000004005E0200000100C70300000200440400000300A00000000400700000000100C70300000200440400000300A00000000100C70300000200440400000300A00000000100B40000000200A00000000100EE0500000200B40000000100110500000100A00000000100A00000000100110500000200EE0500000100C70300000200440400000300A00000000100C70300000200A00000000100440400000200A00000000100C70300000200440400000300A00000000400700000000100A00000000100C70300000200440400000300A00000000100110500000200B40000000300A00000000100110500000200EE0500000300B40000000100BF0400000100BF0400000100BF0400000200A00000000100080500000100440400000200CE0200000100440400000200CE0200000300700000000100440400000200A00000000300CE0200000100C70300000200440400000300A00000000400700000000100A00000000100A00000000100140609008F05010011008F05060019008F050A0061008F05060029005501100079003306150069009F031B006900A70320006900FE052600690009062C0069000800330069001300390079003306400069002002470069002A024C0029008F055200290001015800890056045D003900B906670029007F066F003900C506670031005A067A0031001C04800039003F04890031002B0480002900CA02920069003A06D60029006B03DC004900F001E80008000800F50008000C00FA0008001000FF000800140004010800180009012E000B00DA012E001300E3012E001B000202E1002300F50001012300F50021012300F50041012300F50061012300F50081012300F500A1012300F500C1012300F50001022300F50021022300F50041022300F50061022300F500620075008D009E00A800B200BC00CA00D1000200010000000B05C8010000B204C80100003601CD0100009103CD0100004D02D10100002E05D50102000D00030001000E00030002000F000500010010000500020011000700010012000700020013000900010014000900020017000B00010018000B00020019000D0001001A000D0004800000000000000000000000000000000045060000040000000000000000000000EC0075000000000002000000000000000000000000001C0600000000030002000400020005000200060002000700020008000200350099003500A3003500AD003500B7003500C50000000055496E7433320047657455496E743235360053657455496E74323536003C4D6F64756C653E00696E746572666163654944004765744964546F546F6B656E555249005365744964546F546F6B656E5552490076616C75655F5F00494E6F6E46756E6769626C65546F6B656E4D657461646174610053797374656D2E507269766174652E436F72654C696200696E74657266616365496400546F6B656E496400746F6B656E496400476574417070726F76656400617070726F76656400696400546F6B656E496E746572666163650049537570706F727473496E746572666163650047657442616C616E63650053657442616C616E636500494D657373616765006765745F4D65737361676500494E6F6E46756E6769626C65546F6B656E456E756D657261626C65006765745F4E616D65007365745F4E616D65006E616D650056616C7565547970650043616E4F706572617465006765745F53746174650049536D617274436F6E74726163745374617465004950657273697374656E7453746174650073746174650044656275676761626C6541747472696275746500436F6D70696C6174696F6E52656C61786174696F6E7341747472696275746500496E6465784174747269627574650052756E74696D65436F6D7061746962696C6974794174747269627574650042797465006765745F52657475726E56616C75650076616C756500417070726F76650042616C616E63654F66004F776E65724F6600476574537472696E6700536574537472696E67006765745F4F776E65724F6E6C794D696E74696E67007365745F4F776E65724F6E6C794D696E74696E67006F776E65724F6E6C794D696E74696E67004F776E6572736869705472616E7366657265644C6F67004F776E6572736869705472616E736665725265717565737465644C6F6700417070726F76616C4C6F6700417070726F76616C466F72416C6C4C6F67005472616E736665724C6F670075726900536166655472616E7366657246726F6D496E7465726E616C005472616E73666572496E7465726E616C004C6F67417070726F76616C004765744964546F417070726F76616C005365744964546F417070726F76616C00436C656172417070726F76616C004973417070726F766564466F72416C6C004C6F67417070726F76616C466F72416C6C00536574417070726F76616C466F72416C6C0043616C6C00536D617274436F6E74726163742E646C6C006765745F53796D626F6C007365745F53796D626F6C0073796D626F6C00476574426F6F6C00536574426F6F6C0053797374656D00536166655472616E7366657246726F6D0066726F6D00456E756D00426F6F6C65616E00416464546F6B656E00456E73757265436F6E74726163745265636569766564546F6B656E00494E6F6E46756E6769626C65546F6B656E0052656D6F7665546F6B656E006F705F5375627472616374696F6E006F705F4164646974696F6E004275726E00546F005A65726F00746F00436C61696D4F776E657273686970006765745F53656E646572004C6F675472616E736665720043616E5472616E73666572006765745F4F776E6572007365745F4F776E6572006765745F50656E64696E674F776E6572007365745F50656E64696E674F776E65720053657450656E64696E674F776E657200746F6B656E4F776E6572004765744964546F4F776E6572005365744964546F4F776E65720050726576696F75734F776E65720043757272656E744F776E6572004E65774F776E6572006E65774F776E6572006F776E6572006765745F546F6B656E4964436F756E746572007365745F546F6B656E4964436F756E74657200494E6F6E46756E6769626C65546F6B656E5265636569766572004765744F776E6572546F4F70657261746F72005365744F776E6572546F4F70657261746F720049734F776E65724F724F70657261746F72002E63746F720053797374656D2E446961676E6F737469637300536574537570706F72746564496E74657266616365730053797374656D2E52756E74696D652E436F6D70696C6572536572766963657300446562756767696E674D6F646573006F70657261746F724164647265737300476574416464726573730053657441646472657373006164647265737300537472617469732E536D617274436F6E74726163747300466F726D6174004973436F6E747261637400536D617274436F6E7472616374004F626A656374006F705F496D706C6963697400495472616E73666572526573756C7400536166654D696E7400417373657274004765744964546F417070726F76616C4B6579004765744964546F4F776E65724B657900456E737572654F776E65724F6E6C79006F705F457175616C697479006F705F496E657175616C69747900456E737572654164647265737349734E6F74456D7074790000002D53007500700070006F00720074006500640049006E0074006500720066006100630065003A007B0030007D00001B4900640054006F004F0077006E00650072003A007B0030007D0000214900640054006F0041007000700072006F00760061006C003A007B0030007D000017420061006C0061006E00630065003A007B0030007D00002F4F0077006E006500720054006F004F00700065007200610074006F0072003A007B0030007D003A007B0031007D00000B4F0077006E00650072000019500065006E00640069006E0067004F0077006E006500720000094E0061006D006500000D530079006D0062006F006C00000F5500520049003A007B0030007D0000214F0077006E00650072004F006E006C0079004D0069006E00740069006E006700001D54006F006B0065006E004900640043006F0075006E00740065007200004D5400680065002000660072006F006D00200070006100720061006D00650074006500720020006900730020006E006F007400200074006F006B0065006E0020006F0077006E00650072002E00004F54006800650020007B0030007D0020006100640064007200650073007300200069007300200061006C0072006500610064007900200074006F006B0065006E0020006F0077006E00650072002E00001161007000700072006F007600650064000043430061006C006C006500720020006900730020006E006F00740020006F0077006E006500720020006E006F007200200061007000700072006F007600650064002E000057430061006C006C006500720020006900730020006E006F00740020006F0077006E006500720020006E006F007200200061007000700072006F00760065006400200066006F007200200074006F006B0065006E002E00007143006C00610069006D004F0077006E0065007200730068006900700020006D007500730074002000620065002000630061006C006C0065006400200062007900200074006800650020006E00650077002800700065006E00640069006E006700290020006F0077006E00650072002E00003354006800650020006D006500740068006F00640020006900730020006F0077006E006500720020006F006E006C0079002E0000354F006E004E006F006E00460075006E006700690062006C00650054006F006B0065006E0052006500630065006900760065006400004F4F006E004E006F006E00460075006E006700690062006C00650054006F006B0065006E00520065006300650069007600650064002000630061006C006C0020006600610069006C00650064002E0000494F006E006C007900200074006F006B0065006E0020006F0077006E00650072002000630061006E0020006200750072006E002000740068006500200074006F006B0065006E002E000039540068006500200061006400640072006500730073002000630061006E0020006E006F00740020006200650020007A00650072006F002E0000006B1DBB4E35A7DC4EA54D8D0E5EF93A200004200101080320000105200101111104200012350500020E0E1C042001020E052002010E02052001111D0E062002010E111D05200111190E062002010E11190600030E0E1C1C0420010E0E052002010E0E0520010112210420001245042000111D040701111D07000202111D111D05200201020E04070111190500011119080800021119111911190306111D040701111006300101011E00040A0111100407011114040A0111140407011118040A0111180407011120040A011120080703111D111D111C040A01111C06070211191119040701122505200102111D0B20051225111D0B0E1D1C0B0320001C087CEC85D7BEA7798E040100000004020000000403000000040400000004050000000206080306110C03061119020602042001020906200201110C020520010E1119062001111D1119072002011119111D0620011119111D07200201111D111907200202111D111D08200301111D111D0205200101111D0320000E042001010E0620020111190E03200002042001010204200011190520010111190820040112210E0E020B200401111D111D11191D0509200301111D111D111906200201111D020720021119111D0E0920031119111D0E1D0508200301111D11190E042800111D0328000E0328000204280011190801000800000000001E01000100540216577261704E6F6E457863657074696F6E5468726F7773010801000200000000000000000000000000000000000010000000000000000000000000000000783F00000000000000000000923F0000002000000000000000000000000000000000000000000000843F0000000000000000000000005F436F72446C6C4D61696E006D73636F7265652E646C6C0000000000FF250020001000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000003000000C000000A43F00000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000
```

### CODE REVIEW

**Comment by YakupIpek** [comment]
@mrtpain Can you please review this pr ? We did some improvements based on Certik feedbacks.

File: Testnet/NonFungibleToken/NonFungibleToken/NonFungibleToken.cs
Comment (mrtpain) [inline Testnet/NonFungibleToken/NonFungibleToken/NonFungibleToken.cs:572]: Just a nitpick, set all of them if you're going to set any of them.

File: Testnet/NonFungibleToken/NonFungibleToken/NonFungibleToken.cs
Comment (mrtpain) [inline Testnet/NonFungibleToken/NonFungibleToken/NonFungibleToken.cs:462]: Method looks good but consider adding the comment without the owner parameter since its removed.

File: Testnet/NonFungibleToken/NonFungibleToken/NonFungibleToken.cs
Comment (mrtpain) [inline Testnet/NonFungibleToken/NonFungibleToken/NonFungibleToken.cs:454]: Same comment, method looks good but maybe consider bringing the comment back for method description.

File: Testnet/NonFungibleToken/NonFungibleToken/NonFungibleToken.cs
Comment (mrtpain) [inline Testnet/NonFungibleToken/NonFungibleToken/NonFungibleToken.cs:450]: This can be reduced further using `Assert(CanOperate(tokenOwner) || GetIdToApproval(tokenId) == Message.Sender);`

File: Testnet/NonFungibleToken/NonFungibleToken/NonFungibleToken.cs
Comment (mrtpain) [inline Testnet/NonFungibleToken/NonFungibleToken/NonFungibleToken.cs:371]: Empty address check can be removed probably because `CanTransfer` is called prior which should fail with an empty address.

File: Testnet/NonFungibleToken/NonFungibleToken/NonFungibleToken.cs
Comment (mrtpain) [inline Testnet/NonFungibleToken/NonFungibleToken/NonFungibleToken.cs:299]: Is getting tokenOwner info necessary at all here?

File: Testnet/NonFungibleToken/NonFungibleToken/NonFungibleToken.cs
Comment (mrtpain) [inline Testnet/NonFungibleToken/NonFungibleToken/NonFungibleToken.cs:244]: Same comment here regarding the address not empty check, CanOperate would fail first, think you can remove the extra empty address check here.

**Review by mrtpain — COMMENTED** [review]
Made a first pass through with some suggestions. 

I'll make a second pass through with the full contract context to double check logic between methods and flows don't have any potential conflicts or possible improvements for gas etc.

File: Testnet/NonFungibleToken/NonFungibleToken/NonFungibleToken.cs
Comment (mrtpain) [inline Testnet/NonFungibleToken/NonFungibleToken/NonFungibleToken.cs:108]: Missing comment

File: Testnet/NonFungibleToken/NonFungibleToken/NonFungibleToken.cs
Comment (mrtpain) [inline Testnet/NonFungibleToken/NonFungibleToken/NonFungibleToken.cs:159]: Summary and state param are empty

File: Testnet/NonFungibleToken/NonFungibleToken/NonFungibleToken.cs
Comment (mrtpain) [inline Testnet/NonFungibleToken/NonFungibleToken/NonFungibleToken.cs:161]: Rip this out if this is going up for Testnet/Mainnet vote.

File: Testnet/NonFungibleToken/NonFungibleToken/NonFungibleToken.cs
Comment (YakupIpek) [reply Testnet/NonFungibleToken/NonFungibleToken/NonFungibleToken.cs:299]: Yes because if you remove EnsureAddressIsNotEmpty(tokenOwner); then method will return false 

But method will throws exception in the case of token is not exist. 
So method behavior is changing when the assertion removed.

Comment (mrtpain) [reply Testnet/NonFungibleToken/NonFungibleToken/NonFungibleToken.cs:299]: Shouldn't an approval never be able to be set without an owner? So passing an Id to `GetIdToApproval` should return an address or `Address.Zero` if no approval, so maybe _that_ address should be checked for Empty, not the tokenOwner right? 

Saves some gas not getting the owner when we don't have too if we can just validate the Approval address prior to returning.

File: Testnet/NonFungibleToken/NonFungibleToken/NonFungibleToken.cs
Comment (mrtpain) [inline Testnet/NonFungibleToken/NonFungibleToken/NonFungibleToken.cs:455]: So close haha empty comment

**Comment by mrtpain** [comment]
Only noticed one new thing, other it LGTM. The previous comments being discussed are optional, mainly only gas improvements that I think _could_ be made so I'll leave it up to you.

When you're good with it, if you update the hash and bytecode I can verify that and approve. Nice work this was a big one! 🙌

File: Testnet/NonFungibleToken/NonFungibleToken/NonFungibleToken.cs
Comment (YakupIpek) [reply Testnet/NonFungibleToken/NonFungibleToken/NonFungibleToken.cs:300]: These recommendation change some behavior of method. Both [OpenZeppelin](https://github.com/OpenZeppelin/openzeppelin-contracts/blob/e63b09c9ad3a45484b6dc304e0e99640a9dc3036/contracts/token/ERC721/ERC721.sol#L127) and [0xcert](https://github.com/0xcert/ethereum-erc721/blob/77030ec2a25123d17b13e7ee469f1c956a48451c/src/contracts/tokens/nf-token.sol#L267) implementation is designed this way so i will stick to those implementations.

**Review by YakupIpek — COMMENTED** [review]
@mrtpain I completed all changes now. I reverted some empty address checks because of some reasons. It is how 0xcert how implemented it actually. I prefer to not make change in last minutes because this is safer. 

Let me know if you have any reviews. We need to complete this reviews asap.

**Review by mrtpain — APPROVED** [review]
lgtm

**Comment by StratisIain** [comment]
Whitelisted

**Pull request merged**
